File: Client/Assets/Lua/core/databinding/binding_path_part.lua
------------------------------------------------
--  Copyright © 2013-2020   Hugula: Arpg game Engine
--
--  author pu
------------------------------------------------
local table = table
local string_match = string.match
local string_format = string.format
local type = type
local class = class
local weak_table = weak_table

local self_path = "."

---绑定路径
---源对象值查找和更新
---@class BindingPathPart
local binding_path_part =
    class(
    function(self, binding, path)
        self.binding = binding
        self.path = path
        self._weak_ref = weak_table()
        if path == self_path then
            self.is_self = true
        end

        self.handler = function(sender, args)
            -- Logger.Log(string.format("self=%s,_weak_ref=%s,sender=%s,args=%s", self, self._weak_ref, sender, args))
            -- self:on_property_changed(sender, args)
            local _weak_ref = self._weak_ref
            local listener = _weak_ref["listener"]
            if listener then
                listener(self, sender, args)
            else
                self:unsubscribe()
                Logger.LogError("listener is nil")
            end
        end
    end
)

--- 获取源source的属性
---@overload fun(source:table)
---@param source table
---@return any
local function get_source_val(self, source)
    local val
    if self.is_method == true then
        -- Logger.Log(string_format("get_source_val (source:%s,path:%s())", source, self.path))
        val = source[self.path]()
    else
        -- Logger.Log(string_format("get_source_val(source:%s,path:%s)=%s", source, self.path, val))
        val = source[self.path]
    end
    return val
end

--- 获取源source的属性
---@overload fun(source:table)
---@param source table
---@param val   any
---@return void
local function set_source_val(self, source, val)
    if self.is_method == true then
        local method_args = self.method_args
        if method_args ~= nil and method_args >= 1 then
            source[self.path](val)
        else
            source[self.path]()
        end
    else
        source[self.path] = val
    end
end

--- wm属性改变的时候触发表达式更新
---@overload fun(self:table,property:string)
---@param sender table
---@param property string
local function property_changed(self, sender, property)
    local part = self.next_part or self
    -- Logger.Log("property changed:", self, sender, property,part.path)
    if property and part then
        if
            property == part.path or
                (part.is_indexer and property == string_format("%s[%s]", part.index_name, part.path))
         then --如果匹配 执行绑定函数
            -- Logger.Log("part.binding:Apply(true):", self, sender, property)
            part.binding:Apply(true)
        end
    end
end

--- 取消订阅source改变监听
---@overload fun(self:BindingPathPart)
---@param self BindingPathPart
local function unsubscribe(self)
    local _weak_ref = self._weak_ref
    local source = _weak_ref["source"]
    if source ~= nil and source.PropertyChanged then
        -- Logger.Log("unsubscribe", source, self.path, self.handler)
        source:PropertyChanged("-", self.handler)
    end
    _weak_ref["source"] = nil
    _weak_ref["listener"] = nil
end

--- 取消订阅source改变监听
---@overload fun(self:BindingPathPart,source:INotifyPropertyChanged)
---@param self BindingPathPart
---@param source INotifyPropertyChanged
local function subscribe(self, source)
    if source then
        local _weak_ref = self._weak_ref
        if _weak_ref["source"] ~= source then
            unsubscribe(self)
            _weak_ref["source"] = source
            _weak_ref["listener"] = property_changed
            source:PropertyChanged("+", self.handler)
            -- Logger.Log("subscribe", source, self.path, self.handler)
        end
    end
    -- body
end

local function dispose(self)
    self.handler = nil
    self._weak_ref = nil
end

local function tostring(self)
    -- body
    return string_format("BindingPathPart(path=%s,binding=%s)", self.path, self.binding)
end

binding_path_part.get_source_val = get_source_val
binding_path_part.set_source_val = set_source_val
binding_path_part.subscribe = subscribe
binding_path_part.unsubscribe = unsubscribe
binding_path_part.property_changed = property_changed
binding_path_part.__tostring = tostring

---绑定路径
---源对象值查找和更新
---@class BindingPathPart
---@module binding_path_part.lua
---@field get_source_val function
---@field set_source_val function
---@field subscribe function
---@field unsubscribe function
---@field property_changed function
---@field next_part BindingPathPart
---@field tostring function
BindingPathPart = binding_path_part

File: Client/Assets/Hugula/Databinding/BindingPathPart.cs
using System;
using Hugula.Utils;
using UnityEngine;

namespace Hugula.Databinding
{
    [System.SerializableAttribute]
    public struct BindingPathPartConfig
    {
        [NonSerialized]
        internal PropertyChangedEventHandlerEvent m_NotifyPropertyChanged;

        public string path;
        public BindingPathPartFlags _flags;

        public bool isSelf
        {
            get => (_flags & BindingPathPartFlags.IsSelf) != 0;
            set
            {
                if (value)
                    _flags |= BindingPathPartFlags.IsSelf;
                else
                    _flags &= ~BindingPathPartFlags.IsSelf;
            }
        }

        public bool isIndexer
        {
            get => (_flags & BindingPathPartFlags.IsIndexer) != 0;
            set
            {
                if (value)
                    _flags |= BindingPathPartFlags.IsIndexer;
                else
                    _flags &= ~BindingPathPartFlags.IsIndexer;
            }
        }

        public bool isMethod
        {
            get => (_flags & BindingPathPartFlags.IsMethod) != 0;
            set
            {
                if (value)
                    _flags |= BindingPathPartFlags.IsMethod;
                else
                    _flags &= ~BindingPathPartFlags.IsMethod;
            }
        }

        public bool isExpSetter
        {
            get => (_flags & BindingPathPartFlags.IsExpSetter) != 0;
            set
            {
                if (value)
                    _flags |= BindingPathPartFlags.IsExpSetter;
                else
                    _flags &= ~BindingPathPartFlags.IsExpSetter;
            }
        }

        public bool isExpGetter
        {
            get => (_flags & BindingPathPartFlags.IsExpGetter) != 0;
            set
            {
                if (value)
                    _flags |= BindingPathPartFlags.IsExpGetter;
                else
                    _flags &= ~BindingPathPartFlags.IsExpGetter;
            }
        }

        public bool isGlobal
        {
            get => (_flags & BindingPathPartFlags.IsGlobal) != 0;
            set
            {
                if (value)
                    _flags |= BindingPathPartFlags.IsGlobal;
                else
                    _flags &= ~BindingPathPartFlags.IsGlobal;
            }
        }

        /// <summary>
        /// 是否是表达式
        /// </summary>
        public bool isExpression
        {
            get => (_flags & BindingPathPartFlags.IsExpSetter) != 0 || (_flags & BindingPathPartFlags.IsExpGetter) != 0;
        }

#if UNITY_EDITOR
        public override string ToString()
        {
            return string.Format("BindingPathPartConfig(path={0},isSelf={1},isIndexer={2},isMethod={3},isExpSetter={4},isExpGetter={5},) m_NotifyPropertyChanged={6}", this.path, isSelf, isIndexer, isMethod, isExpSetter, isExpGetter, m_NotifyPropertyChanged);
        }
#endif

        public void Dispose()
        {
            path = null;
            m_NotifyPropertyChanged = null;
            isIndexer = false;
            isExpSetter = false;
            isSelf = false;
        }
    }

    /***
    public class BindingPathPart  //65B
    {


        public BindingPathPart nextPart { get; set; }

        public string path;//{ get; internal set; }

        private BindingPathPartFlags _flags = BindingPathPartFlags.None;


        // public string indexerName { get; set; } //indexerName可以删除

        // public bool isIndexer { get; internal set; }

        // public bool isSelf { get; private set; }

        // //表示方法
        // public bool isMethod { get; internal set; }
        public bool isIndexer
        {
            get => (_flags & BindingPathPartFlags.IsIndexer) != 0;
            internal set
            {
                if (value)
                    _flags |= BindingPathPartFlags.IsIndexer;
                else
                    _flags &= ~BindingPathPartFlags.IsIndexer;
            }
        }

        public bool isSelf
        {
            get => (_flags & BindingPathPartFlags.IsSelf) != 0;
            private set
            {
                if (value)
                    _flags |= BindingPathPartFlags.IsSelf;
                else
                    _flags &= ~BindingPathPartFlags.IsSelf;
            }
        }

        public bool isMethod
        {
            get => (_flags & BindingPathPartFlags.IsMethod) != 0;
            internal set
            {
                if (value)
                    _flags |= BindingPathPartFlags.IsMethod;
                else
                    _flags &= ~BindingPathPartFlags.IsMethod;
            }
        }

        // WeakReference<object> m_SourceWeak = new WeakReference<object>(null);

        //当前节点source对象缓存结果加快访问
        public object source;
        // {
        //     get
        //     {
        //         if (m_SourceWeak.TryGetTarget(out var target))
        //             return target;
        //         return null;
        //     }
        //     set
        //     {
        //         m_SourceWeak.SetTarget(value);
        //     }
        // }

        //INotifyPropertyChanged or PropertyChangedEventHandlerEvent
        // WeakReference<object> m_NotifyPropertyChangedWeak = new WeakReference<object>(null);

        //监听的事件
        PropertyChangedEventHandlerEvent m_NotifyPropertyChanged;
        // {
        //     get
        //     {
        //         if (m_NotifyPropertyChangedWeak.TryGetTarget(out var target))
        //             return target;
        //         return null;
        //     }
        //     set
        //     {
        //         m_NotifyPropertyChangedWeak.SetTarget(value);
        //     }
        // }

        internal Binding m_Binding;

        // readonly PropertyChangedEventHandler m_ChangeHandler;

        ~BindingPathPart()
        {

        }

        /// <summary>
        /// for pool
        /// </summary>
        public BindingPathPart()
        {
            // m_ChangeHandler = PropertyChanged;
        }

        /// <summary>
        /// 初始化
        /// </summary>
        public void Initialize(Binding binding, string path, bool isIndexer = false, bool isMethod = false)
        {
            this.m_Binding = binding;
            isSelf = path == Binding.SelfPath;
            this.path = path;
            this.isIndexer = isIndexer;
            this.isMethod = isMethod;

            // m_ChangeHandler = PropertyChanged;
        }

        public void Initialize(Binding binding, BindingPathPartConfig config)
        {
            this.m_Binding = binding;
            this.isSelf = config.isSelf;
            this.path = config.path;
            this.isIndexer = config.isIndexer;
            this.isMethod = config.isMethod;
        }


        /// <summary>
        /// 订阅 hander必须是 PropertyChangedEventHandlerEvent 
        /// </summary>
        /// <param name="handler"></param>
        public void Subscribe(PropertyChangedEventHandlerEvent handler)
        {
            if (ReferenceEquals(m_NotifyPropertyChanged, handler))
                return;

            Unsubscribe(false);

            BindingPathPart part = this;
            var propertName = part.path;
            handler.Add(PropertyChanged, propertName);
            m_NotifyPropertyChanged = handler;
        }


        public void Unsubscribe(bool sourceRelease = true)
        {
            if (m_NotifyPropertyChanged == null) return;

            BindingPathPart part = this;
            var propertName = part.path;
            m_NotifyPropertyChanged.Remove(PropertyChanged, propertName);
            m_NotifyPropertyChanged = null;//

            if (sourceRelease)
                source = null;
        }

        public void PropertyChanged(object sender, string propertyName)
        {
            if (m_Binding == null)
                return;

            BindingPathPart part = this;
            if (!string.IsNullOrEmpty(propertyName))
            {
                if (propertyName != part.path)
                {
                    return;
                }
            }
            // m_Binding.OnSourceChanged(this);

        }

        public bool TryGetValue(object source, out object value)
        {
            value = source;
            if (value != null)
            {
                if (source is XLua.LuaTable lua)
                {
                    lua.TryGet<string, object>(path, out value);
                }
                else
                {
                    value = ExpressionUtility.GetSourceInvoke(value, this);
                }

                return true;
            }
            return false;
        }

        public void Dispose()
        {
            Unsubscribe();
            // indexerName = null;
            isIndexer = false;
            isMethod = false;
            isSelf = false;
            path = null;

            source = null;
            nextPart = null;
            m_Binding = null;
            m_NotifyPropertyChanged = null;

        }

        public void ReleaseToPool()
        {
            m_PoolBindingPathPart.Release(this);
        }

        public override string ToString()
        {
            return string.Format("BindingPathPart(path={0},isIndexer={1},isMethod={2},isSelf={3},source={4})", this.path, isIndexer, isMethod, isSelf, source);
        }

        #region  pool

        public static BindingPathPart Get()
        {
            return m_PoolBindingPathPart.Get();
        }

        const int capacity = 4096;
        const int initial = 2048;

        private static ObjectPool<BindingPathPart> m_PoolBindingPathPart = new ObjectPool<BindingPathPart>(null, ActionOnRelease, capacity, initial);

        private static void ActionOnRelease(BindingPathPart bpPart)
        {
            bpPart.Dispose();
        }

        #endregion
    }

        ***/
    [Flags]
    public enum BindingPathPartFlags : byte
    {
        None = 0,
        IsSelf = 1 << 0,
        IsIndexer = 1 << 1,
        IsMethod = 1 << 2,
        IsExpSetter = 1 << 3,

        IsExpGetter = 1 << 4,
        /// <summary>
        /// 表达式上下文是否全局
        /// </summary>
        IsGlobal = 1 << 5,

    }

}